operator/v1/types_ingress.go

```diff
@@ -1392,6 +1392,38 @@ type IngressControllerTuningOptions struct {
 	// +kubebuilder:validation:Format=duration
 	// +optional
 	HealthCheckInterval *metav1.Duration `json:"healthCheckInterval,omitempty"`
+
+	// maxConnections defines the maximum number of simultaneous
+	// connections that can be established per HAProxy process.
+	// Increasing this value allows each ingress controller pod to
+	// handle more connections but at the cost of additional
+	// system resources being consumed.
+	//
+	// Permitted values are:
+	//
+	//  empty
+	//  0
+	//  -1
+	//  2000-2000000
+	//
+	// If this field is empty or 0, the IngressController will use
+	// the default value of 20000, but the default is subject to
+	// change in future releases.
+	//
+	// If the value is -1 then HAProxy will dynamically compute a
+	// maximum value based on the available ulimits in the running
+	// container. Selecting -1 (i.e., auto) will result in a large
+	// value being computed (~520000 on OpenShift >=4.10 clusters)
+	// and therefore each HAProxy process will incur significant
+	// memory usage compared to the current default of 20000.
```

Comment:
How does the operator manage this?  Will it automatically "fix" the memory  usage request?  I would expect the operator to be responsible for making a pod that is likely to function.

Reply:
> So how do we manage the potentially unbounded growth of haproxy processes today? We have an annotation ingress.operator.openshift.io/hard-stop-after: <duration> that will kill hard any extant haproxy process that runs on beyond <duration>.

What is this annotation on?  We avoid annotations in favor of fields one APIs that we own.

How will a user know that they need to set this hard-stop-after and how will they know they are under memory pressure when they use this auto value?

Is auto truly required to satisfy the use-case?

Reply:
> What is this annotation on? We avoid annotations in favor of fields one APIs that we own.

On an individual ingresscontroller or the ingress config: https://github.com/openshift/cluster-ingress-operator/pull/522

This change was also backported to OCP 4.5. We chose an annotation because it could be backported; there were was no other mitigation in older OCP releases beyond manually identifying and killing haproxy processes. It still could (should?) become an API field.

> How will a user know that they need to set this hard-stop-after and how will they know they are under memory pressure when they use this auto value?

They have to monitor prometheus metrics for the router pods. If you go with auto the same monitoring still applies.

> Is auto truly required to satisfy the use-case?

If you want a large value it saves guessing. If you manually select a value and the pod is migrated to another node there's no guarantee that the new node has identical ulimit settings. If you use auto it would adapt without requiring manual intervention.

Reply:
> We chose an annotation because it could be backported; there were was no other mitigation in older OCP releases beyond manually identifying and killing haproxy processes. It still could (should?) become an API field.

It could always have been a new API field.

> They have to monitor prometheus metrics for the router pods. If you go with auto the same monitoring still applies.

What specifically?  Imagine the help page being written.

> If you want a large value it saves guessing. If you manually select a value and the pod is migrated to another node there's no guarantee that the new node has identical ulimit settings. If you use auto it would adapt without requiring manual intervention.

Explaining the benefit of this over choosing a value is useful.

Reply:
Feedback addressed in [latest push](https://github.com/openshift/api/commit/0c7ae02663d4de5dd38267e1892725e47f25cbee).